std/random.d

```diff
@@ -630,93 +708,78 @@ Parameters for the generator.
-            upperMask = ~((cast(UIntType) 1u <<
-                           (UIntType.sizeof * 8 - (w - r))) - 1),
-            lowerMask = (cast(UIntType) 1u << r) - 1;
-        static immutable UIntType[2] mag01 = [0x0UL, a];
-
-        ulong y = void;
-
-        if (mti >= n)
-        {
-            /* generate N words at one time */
-
-            int kk = 0;
-            const limit1 = n - m;
-            for (; kk < limit1; ++kk)
-            {
-                y = (mt[kk] & upperMask)|(mt[kk + 1] & lowerMask);
-                mt[kk] = cast(UIntType) (mt[kk + m] ^ (y >> 1)
-                        ^ mag01[cast(UIntType) y & 0x1U]);
-            }
-            const limit2 = n - 1;
-            for (; kk < limit2; ++kk)
-            {
-                y = (mt[kk] & upperMask)|(mt[kk + 1] & lowerMask);
-                mt[kk] = cast(UIntType) (mt[kk + (m -n)] ^ (y >> 1)
-                                         ^ mag01[cast(UIntType) y & 0x1U]);
-            }
-            y = (mt[n -1] & upperMask)|(mt[0] & lowerMask);
-            mt[n - 1] = cast(UIntType) (mt[m - 1] ^ (y >> 1)
-                                        ^ mag01[cast(UIntType) y & 0x1U]);
-
-            mti = 0;
-        }
-
-        y = mt[mti++];
-
-        /* Tempering */
-        y ^= (y >> temperingU);
-        y ^= (y << temperingS) & temperingB;
-        y ^= (y << temperingT) & temperingC;
-        y ^= (y >> temperingL);
-
-        _y = cast(UIntType) y;
+        return this.state.y;
     }
 
-/**
-   Returns the current random value.
- */
-    @property UIntType front() @safe pure nothrow @nogc
+    /// Range primitive: advance the generator state
+    /// to get the next random variate
+    void popFront() @safe pure nothrow @nogc
     {
-        if (mti == size_t.max) seed();
-        return _y;
+        this.popFrontImpl(this.state);
+    }
+
+    /// Internal implementation of `popFront()`, which
+    /// can be used with an arbitrary `State` instance
+    private static void popFrontImpl(ref State mtState)
+    {
+        // This function blends two nominally independent
+        // processes: (i) calculation of the next random
+        // variate `mtState.y` from the cached previous
+        // `data` entry `z`, and (ii) updating the value
+        // of `data[index]` and `mtState.z` and advancing
+        // the `index` value to the next in sequence.
+        //
+        // By interweaving the steps involved in these
+        // procedures, rather than performing each of
+        // them separately in sequence, the variables
+        // are kept 'hot' in CPU registers, allowing
+        // for significantly faster performance.
+        sizediff_t index = cast(size_t)mtState.index;
+        sizediff_t next = index - 1;
+        if (next < 0)
+            next = n - 1;
+        auto z = mtState.z;
+        sizediff_t conj = index - m;
+        if (conj < 0)
+            conj = index - m + n;
+        static if (d == UIntType.max)
+            z ^= (z >> u);
```

Comment:
Interesting that code coverage analysis suggests this path is never taken; I would have expected it to be normal for the 32-bit `Mt19937`.